packages/channel-client/tests/channel-client.test.ts

```diff
@@ -0,0 +1,198 @@
+    ).build();
+
+    states['running'] = ChannelResultBuilder.from(states['proposed'])
+      .setStatus('running')
+      .setTurnNum('3')
+      .build();
+
+    states['updated_app_data'] = ChannelResultBuilder.from(states['running'])
+      .setAppData(UPDATED_APP_DATA)
+      .setTurnNum('4')
+      .build();
+
+    states['closed'] = ChannelResultBuilder.from(states['running'])
+      .setStatus('closed')
+      .setTurnNum('5')
+      .build();
+  });
+
+  function setupProvider(provider: FakeChannelProvider, playerIndex: number, addresses: Addresses) {
+    provider.setAddress(addresses.self);
+    provider.updatePlayerIndex(playerIndex);
+    provider.opponentAddress = addresses.opponent;
+  }
+
+  beforeEach(() => {
+    providerA = new FakeChannelProvider();
+    providerB = new FakeChannelProvider();
+
+    clientA = new ChannelClient(providerA);
+    clientB = new ChannelClient(providerB);
+
+    setupProvider(providerA, 0, {
+      self: participantA.participantId,
+      opponent: participantB.participantId
+    });
+
+    setupProvider(providerB, 1, {
+      self: participantB.participantId,
+      opponent: participantA.participantId
+    });
+
+    // This setup simulates the message being received from A's wallet
+    // and "queued" by A's app to be sent to the opponent (handled by
+    // A's channel client, which then is "dequeued" on A's channel client
+    // and sent to B's app (handled by B's channel client here) and
+    // pushed from B's app to B's wallet.
+    // The de/queuing described above is effectively faked by explicitly passing
+    // the messages between the clients.
+    clientA.onMessageQueued(async (message: Message<ChannelResult>) => {
+      await clientB.pushMessage(message);
+    });
+
+    clientB.onMessageQueued(async (message: Message<ChannelResult>) => {
+      await clientA.pushMessage(message);
+    });
+
+    clientB.onChannelProposed((result: ChannelResult) => {
+      clientBEventEmitter.emit('ChannelProposed', result);
+    });
+  });
+
+  describe('creates a channel', () => {
+    let proposalMessage: Message<ChannelResult>;
+
+    it('client A produces the right channel result', async () => {
+      const clientAChannelState = await clientA.createChannel(
+        participants,
+        allocations,
+        APP_DEFINITION,
+        APP_DATA
+      );
+      expect(clientAChannelState).toEqual(states['proposed']);
+
+      proposalMessage = {
+        sender: await clientA.getAddress(),
+        recipient: await clientB.getAddress(),
+        data: clientAChannelState
+      };
+    });
+
+    it('client B gets proposal', async () => {
+      setProviderStates([providerA, providerB], states['proposed']);
+
+      return new Promise(resolve => {
+        clientBEventEmitter.once('ChannelProposed', (result: ChannelResult) => {
+          expect(providerB.latestState).toEqual(states['proposed']);
+          resolve();
+        });
+
+        clientB.pushMessage(proposalMessage);
```

Comment:
Is it a bit unnatural to have an expectation declared *before* the function call that should trigger it to be met?

Or have I misunderstood what is happening here?

Reply:
The exception is only declared to be expected in the context of that test block, but what actually triggers it is the code that follows it.

Setting the expect after the trigger won't call the intended expect code, as the `once` hook gets registered after the trigger and it essentially turns into a race.